text/0554-deprecate-getwithdefault.md

```diff
@@ -0,0 +1,71 @@
+- Start Date: 2019-11-08
+- Relevant Team(s): Ember.js
+- RFC PR: https://github.com/emberjs/rfcs/pull/554
+- Tracking: (leave this empty)
+
+# Deprecate getWithDefault
+
+## Summary
+
+Deprecate support for `getWithDefault` in Ember's Object module (@ember/object) because its expected behaviour is confusing to Ember developers.
+
+## Motivation
+
+The problem with `getWithDefault` is that its behaviour is confusing to Ember developers. The API will only return the default value when the value of the property retrieved is `undefined`. This behaviour is often overlooked when using the function where a developer might expect that `null` or other _falsy_ values will also return the default value.
+
+Given the JavaScript language will soon (currently in Stage 3) give us the appropriate tool for this use case using the [Nullish Coalescing Operator `??`](https://github.com/tc39/proposal-nullish-coalescing), we can deprecate usage of `getWithDefault` and use that instead.
+
+## Transition Path
+
+Ember will start logging deprecation messages for `getWithDefault` usage. 
+
+We can codemod our current usage of `getWithDefault` with the equivalent behaviour using plain JavaScript. The migration guide will cover this example:
+
+Before:
+
+```js
+let result = getWithDefault(obj, ‘some.key', defaultValue);
+```
+
+After:
+
+```js
+let result = get(obj, 'some.key');
+if (result === undefined) {
+  result = defaultValue;
+}
+```
```

Comment:
```js
let result = get(obj, 'some.key') || defaultValue;
```

Is this an alternative option?  Also, should the codemod require `get` or can we use plain property accessors?

Reply:
I think it makes sense to use `get` to minimize behaviour differences

Reply:
Codemodding paths and interpolated paths will be significantly harder as well:

```js
getWithDefault(obj, 'foo.bar');
getWithDefault(obj, `foo.${key}`);
getWithDefault(obj, path);
```

I'd also suggest to codemod towards `get` and then potentially run another codemod from `get` towards native accessors in the future.

Reply:
FWIW, there is already a codemod for `get()` to native: https://github.com/ember-codemods/es5-getter-ember-codemod. (at least for simple properties, no paths yet)

And someone even suggested covering `getWithDefault()` already: https://github.com/ember-codemods/es5-getter-ember-codemod/issues/1 😝